File: tree/103.binary-tree-zigzag-level-order-traversal.py
#
# @lc app=leetcode id=103 lang=python3
#
# [103] Binary Tree Zigzag Level Order Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        res = []
        queue = []
        cnt = 1

        if root:
            queue.append(root)

        while queue:
            level = []
            n = len(queue)
            for _ in range(n):
                node = queue.pop(0)
                level.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            if cnt % 2 == 1:
                res.append(level)
            else:
                level.reverse()
                res.append(level)
            cnt += 1

        return res

# @lc code=end

File: 103.binary-tree-zigzag-level-order-traversal.go
package leetcode

// zigzagLevelOrder TODO
// 103. 二叉树的锯齿形层次遍历
func zigzagLevelOrder(root *TreeNode) [][]int {
	ret := [][]int{}
	if root == nil {
		return ret
	}
	inverted := false
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		n := len(queue)
		var layer []int
		for j := 0; j < n; j++ {
			node := queue[j]
			layer = append(layer, node.Val)
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		if inverted {
			for i := 0; i < len(layer)/2; i++ {
				layer[i], layer[len(layer)-i-1] = layer[len(layer)-i-1], layer[i]
			}
		}
		inverted = !inverted
		ret = append(ret, layer)
		queue = queue[n:]
	}
	return ret
}
